File: 3_longest_substring.rb
# @param {String} s
# @return {Integer}
def length_of_longest_substring(s)
  maxlength = 0
  temp = [0, 0] # [start_pos, length]

  bytes = s.bytes
  bytes.each_index do |i|
    dup_index = bytes[temp.first, temp.last].index(bytes[i])
    if dup_index
      maxlength = temp.last if temp.last > maxlength

      temp[0] += dup_index + 1
      temp[1] -= dup_index
    else
      temp[1] += 1
    end
  end
  temp.last > maxlength ? temp.last : maxlength
end

File: 3_longest_substring.js
/**
 * @param {string} s
 * @return {number}
 */
var lengthOfLongestSubstring = function(s) {
  if (s.length == 0) return 0;
  
  var map = {},
      length = 0,
      maxlength = 0;
  
  for (i = 0; i < s.length; i++) {
      char = s[i];
      if (char in map) {
          newPos = map[char] + 1;
          lasti = i;
          i = newPos;
          map = {};
          map[s[i]] = i;
          maxlength = Math.max(length, maxlength);
          length = 1;
      } else {
          map[char] = i;
          length++;
      }
  }
  maxlength = Math.max(length, maxlength);
  
  return maxlength;
};
